### PR TITLE
New sample: LivePerson connector

### PR DESCRIPTION
**Pull request by arturl**
The PR has a new sample and up update to the EventFactory (already in SDK [master])

### CODE REVIEW

File: experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/Bots/LPProxyBot.cs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+using LivePersonConnector;
+using Microsoft.Bot.Builder;
+using Microsoft.Bot.Connector;
+using Microsoft.Bot.Schema;
+using Microsoft.Extensions.Configuration;
+using Newtonsoft.Json.Linq;
+
+namespace LPProxyBot.Bots
+{
+    public class LPProxyBot : ActivityHandler
```
Comment (carlosscastro) [inline experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/Bots/LPProxyBot.cs]: I am not sure I love the name. LivePersonProxyBot sounds way more descriptive. Thoughts?

File: experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LPCredentialsProvider.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using LivePersonConnector;
+using Microsoft.Extensions.Configuration;
+
+namespace LPProxyBot
+{
+    public class LPCredentialsProvider : ICredentialsProvider
```
Comment (carlosscastro) [inline experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LPCredentialsProvider.cs]: try to not use acronyms in class and module names please

File: experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LPCredentialsProvider.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        public LPCredentialsProvider(IConfiguration configuration)
+        {
+            LpAccount = configuration["LivePersonAccount"];
+#if RELEASE
```
Comment (carlosscastro) [inline experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LPCredentialsProvider.cs]: Do we need conditional compilation here? People can use the appsettings.json transformation perhaps?

File: experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/ConversationMap.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+using System;
+using System.Collections.Concurrent;
+using Microsoft.Bot.Schema;
+
+namespace LivePersonConnector
+{
+    public class ConversationRecord
+    {
+        public ConversationReference ConversationReference;
+        public bool IsAcked = false;
```
Comment (carlosscastro) [inline experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/ConversationMap.cs]: IsAcknoledged

Also if it is public ideally make it a property

Comment (carlosscastro) [reply experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/ConversationMap.cs]: Property comment applies to entire class

File: experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/HandoffMiddleware.cs [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+                }
+                catch { }
+            }
+
+            turnContext.OnSendActivities(async (sendTurnContext, activities, nextSend) =>
+            {
+                // Handle any escalation events, and let them propagate through the pipeline
+                // This is useful for debugging with the Emulator
+                var handoffEvents = activities.Where(activity =>
+                    activity.Type == ActivityTypes.Event && activity.Name == HandoffEventNames.InitiateHandoff);
```
Comment (carlosscastro) [inline experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/HandoffMiddleware.cs]: Are we case sensitive for activity names? Just checking

File: experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/HandoffMiddleware.cs [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+                if (handoffEvents.Count() == 1)
+                {
+                    var handoffEvent = handoffEvents.First();
+                    conversationData.EscalationRecord = await Escalate(sendTurnContext, handoffEvent);
```
Comment (carlosscastro) [inline experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/HandoffMiddleware.cs]: SHould we stick to the naming convention we have across la SDK where all async method names end with the Async keyword, i.e. rename this method to EscalateAsync

File: experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/HandoffMiddleware.cs [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+                    await _conversationState.SaveChangesAsync(turnContext);
+                }
+
+                // run full pipeline
+                var responses = await nextSend().ConfigureAwait(false);
```
Comment (carlosscastro) [inline experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/HandoffMiddleware.cs]: Note that the sample is not consistent with the use of .ConfigureAwait(false). All sample should be consistent in all async calls

File: experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/LivePersonConnector.cs [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Bot.Builder;
+using Microsoft.Bot.Schema;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace LivePersonConnector
+{
+    public interface ICredentialsProvider
```
Comment (carlosscastro) [inline experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/LivePersonConnector.cs]: There is an ICredentialProvider core interafce in the SDK and folks might get confused. Also, given that these are properties, do we need an interface at all? Will you or anybody provide different implementations of a property-only class without behavior?

File: experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/LivePersonConnector.cs [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+    {
+        string LpAccount { get; }
+        string LpAppId { get; }
+        string LpAppSecret { get; }
+        string MsAppId { get; }
+    }
+
+    internal class LivePersonConversationRecord
+    {
+        public string ConversationId;
```
Comment (carlosscastro) [inline experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/LivePersonConnector.cs]: Prefer properties over public member variables

File: experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/LivePersonConnector.cs [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+        public string MsgDomain;
```
Comment (carlosscastro) [inline experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/LivePersonConnector.cs]: Acronyms are not ideal for public names in the context of samples

File: experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/LivePersonConnector.cs [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+        public string AppJWT;
+        public string ConsumerJWS;
```
Comment (carlosscastro) [inline experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/LivePersonConnector.cs]: What is a JWS? Maybe some comments to explain what the fields in this class mean? that is, if users need to understand.

File: experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/LivePersonConnector.cs [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+    }
+
+    static class LivePersonConnector
+    {
+        static public async Task<LivePersonConversationRecord> EscalateToAgent(ITurnContext turnContext, IEventActivity handoffEvent, string account, string clientId, string clientSecret, ConversationMap conversationMap)
+        {
+            var sentinelDomain = await GetDomain(account, "sentinel");
+            var appJWT = await GetAppJWT(account, sentinelDomain, clientId, clientSecret);
```
Comment (carlosscastro) [inline experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/LivePersonConnector.cs]: consider adding Async at the end of the asybc method. GetAppJwtAsync?

File: experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/LivePersonConnector.cs [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+            var userName = turnContext.Activity.From.Name;
+
+            var idpDomain = await GetDomain(account, "idp");
+            var consumerJWS = await GetConsumerJWS(account, idpDomain, appJWT, consumer);
+
+            // This can be null:
+            var skill = (handoffEvent.Value as JObject)?.Value<string>("Skill");
+
+            var msgDomain = await GetDomain(account, "asyncMessagingEnt");
+            var conversations = new Conversation[] {
```
Comment (carlosscastro) [inline experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/LivePersonConnector.cs]: Is this a place where developers would change this code?

File: experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/Bots/LPProxyBot.cs (hunk 1 above)
Comment (arturl) [reply experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/Bots/LPProxyBot.cs]: Agree. Fixing this and other abbreviations.

File: experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/HandoffMiddleware.cs (hunk 5 above)
Comment (arturl) [reply experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/HandoffMiddleware.cs]: yes, case sensitive

File: experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LPCredentialsProvider.cs (hunk 3 above)
Comment (arturl) [reply experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LPCredentialsProvider.cs]: Removing

File: experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/LivePersonConnector.cs (hunk 12 above)
Comment (arturl) [reply experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/LivePersonConnector.cs]: Agreed

File: experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/LivePersonConnector.cs (hunk 8 above)
Comment (arturl) [reply experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/LivePersonConnector.cs]: Renamed. This is used by the connector but implemented by the user for separation of concerns. A struct would also work but the user (bot) still has to instantiate and pass into the connector, so it's not really much simpler.

File: experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/LivePersonConnector.cs (hunk 9 above)
Comment (arturl) [reply experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/LivePersonConnector.cs]: done, everywhere

File: experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/LivePersonConnector.cs (hunk 11 above)
Comment (arturl) [reply experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/LivePersonConnector.cs]: added comment

File: experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/LivePersonConnector.cs (hunk 13 above)
Comment (arturl) [reply experimental/handoff-library/csharp_dotnetcore/samples/LPProxyBot/LiverPersonConnector/LivePersonConnector.cs]: Good point, I made it part of the context. Now it's coming from the bot, developers don't have to change the connector code.

**Pull request merged**